### PR TITLE
Don't expose TextFieldMapper subfields

### LINKED ISSUE #63446: Text fields should not expose internal field types
When `text` fields has `index_prefixes` or `index_phrases` enabled, they create subfields of the form `field._index_prefixes`. These fields are only meant for internal optimizations and shouldn't be searched on directly.

However the text mapper exposes these hidden fields as if they were configured as multi-fields. This means that field caps includes the hidden fields, as when requesting all fields through `fields=*`. The search 'fields' option will also match against them, which can add duplication to the output.

Perhaps we should stop exposing these fields, keeping them an implementation detail. Note that the same consideration applies to `search_as_you_type`.

### LINKED ISSUE #64585: Exists query fail on text field with sub-fields option
**Elasticsearch version** (`bin/elasticsearch --version`):

7.9.2

**Plugins installed**: []

**Description of the problem including expected versus actual behavior**:

Text fields that activate `index_prefixes` or `index_phrases` break `exists` query that targets
the parent object. So currently if a field `foo.bar` is defined as a text  field with `index_prefixes` option, 
any `exists` query  on `foo` throws an `UOE`:

```
org.elasticsearch.index.mapper.TextFieldMapper$PrefixFieldType.existsQuery(TextFieldMapper.java:488)
  at org.elasticsearch.index.query.ExistsQueryBuilder.newObjectFieldExistsQuery(ExistsQueryBuilder.java:214)
  at org.elasticsearch.index.query.ExistsQueryBuilder.newFieldExistsQuery(ExistsQueryBuilder.java:202)
  at org.elasticsearch.index.query.ExistsQueryBuilder.newFilter(ExistsQueryBuilder.java:164)
  at org.elasticsearch.index.search.QueryStringQueryParser.existsQuery(QueryStringQueryParser.java:636)
  at org.elasticsearch.index.search.QueryStringQueryParser.getWildcardQuery(QueryStringQueryParser.java:647) 
```

**Steps to reproduce**:

```yml

PUT test
{
    "mappings": {
        "properties": {
            "foo": {
                "properties": {
                    "bar": {
                        "type": "text",
                        "index_prefixes": {}
                    }
                }
            }
        }
    }
}

POST test/_doc
{
  "foo.bar": "1"
}

POST test/_search?q=foo:*
```

Relates #63446

### PR DESCRIPTION
**Pull request by romseygeek**
TextFieldMapper can optionally index data into subfields for accelerated
prefix and phrase queries.  Currently, these subfields are implemented
as FieldMappers in their own right, made available via TextFieldMapper's
iterator() method and with their own standalone MappedFieldType objects.

This has the disadvantage that these subfields are directly available for
searching, and appear in APIs such as field caps.  In addition, because
exists queries are not implemented on them, an exists query against an
object which contains a text field with one of the subfields enabled can
throw an error (see #63585).

This commit reworks the subfields so that they are no longer implemented
as FieldMappers, and are no longer exposed to classes outside
TextFieldMapper either as MappedFieldTypes or as FieldMappers.  The
parent TextFieldMapper handles indexing and analyzer registration, 
PhraseFieldType is removed entirely, and PrefixFieldType is retained as
a private implementation for fast prefix queries but is unavailable for
querying directly.

Fixes #63585
Closes #63446
Closes #64585

### CODE REVIEW

**Comment by romseygeek** [comment]
I've marked this as `>breaking` as it means that you'll no longer be able to search directly on the indexed subfields, but I think that's a good thing so I don't think we need to worry about backwards compatibility here.

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 1]
```diff
@@ -330,7 +329,7 @@ private TextFieldType buildFieldType(FieldType fieldType, BuilderContext context
             return ft;
         }
 
-        private PrefixFieldMapper buildPrefixMapper(BuilderContext context, FieldType fieldType, TextFieldType tft) {
+        private SubFieldInfo buildPrefixMapper(BuilderContext context, FieldType fieldType, TextFieldType tft) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:332]: buildPrefixMapper -> buildPrefixInfo

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 2]
```diff
@@ -407,12 +405,18 @@ private PhraseFieldMapper buildPhraseMapper(FieldType fieldType, TextFieldType p
         public TextFieldMapper build(BuilderContext context) {
             FieldType fieldType = TextParams.buildFieldType(index, store, indexOptions, norms, termVectors);
             TextFieldType tft = buildFieldType(fieldType, context);
-            PhraseFieldMapper phraseFieldMapper = buildPhraseMapper(fieldType, tft);
-            PrefixFieldMapper prefixFieldMapper = buildPrefixMapper(context, fieldType, tft);
+            SubFieldInfo phraseFieldMapper = buildPhraseInfo(fieldType, tft);
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:408]: We could switch to the `*Info` naming here too,

**Review by jtibshirani — APPROVED** [review]
This makes sense to me!

**Pull request merged**